photon-api/src/main/scala/com/linkedin/photon/ml/data/LocalDataset.scala

```diff
@@ -188,61 +188,69 @@ object LocalDataset {
   protected[ml] def computePearsonCorrelationScore(
       labelAndFeatures: Array[(Double, Vector[Double])]): Map[Int, Double] = {
 
-    val featureLabelProductSums = mutable.Map[Int, Double]()
-    val featureFirstOrderSums = mutable.Map[Int, Double]()
-    val featureSecondOrderSums = mutable.Map[Int, Double]()
-    var labelFirstOrderSum = 0.0
-    var labelSecondOrderSum = 0.0
-    var numSamples = 0
+    val featureMeans = mutable.Map[Int, Double]()
+    var featureUnscaledVars = mutable.Map[Int, Double]()
+    var labelMean = 0.0
+    var labelUnscaledVar = 0.0
+    val unscaledCovariances = mutable.Map[Int, Double]()
     var interceptAdded = false
+    var numSamples = 0
 
     labelAndFeatures.foreach { case (label, features) =>
       numSamples += 1
-      labelFirstOrderSum += label
-      labelSecondOrderSum += label * label
+
+      val deltaLabel = label - labelMean
+      labelMean += deltaLabel / numSamples
+      labelUnscaledVar += deltaLabel * (label - labelMean)
+
       // Note that, if there is duplicated keys in the feature vector, then the following Pearson correlation scores
       // calculation will screw up
       features.activeIterator.foreach { case (key, value) =>
-        featureFirstOrderSums.update(key, featureFirstOrderSums.getOrElse(key, 0.0) + value)
-        featureSecondOrderSums.update(key, featureSecondOrderSums.getOrElse(key, 0.0) + value * value)
-        featureLabelProductSums.update(key, featureLabelProductSums.getOrElse(key, 0.0) + value * label)
+        val prevFeatureMean = featureMeans.getOrElse(key, 0.0)
+        val deltaFeature = value - prevFeatureMean
+        val featureMean = prevFeatureMean + deltaFeature / numSamples
+
+        val prevFeatureUnscaledVar = featureUnscaledVars.getOrElse(key, 0.0)
+        val featureUnscaledVar = prevFeatureUnscaledVar + (value - prevFeatureMean) * (value - featureMean)
+
+        val prevCovariance = unscaledCovariances.getOrElse(key, 0.0)
+        val unscaledCovariance = prevCovariance + deltaFeature * deltaLabel * (numSamples - 1) / numSamples
+
+        featureMeans.update(key, featureMean)
+        featureUnscaledVars.update(key, featureUnscaledVar)
+        unscaledCovariances.update(key, unscaledCovariance)
       }
     }
 
-    featureFirstOrderSums
+    val labelStd = math.sqrt(labelUnscaledVar / numSamples)
```

Comment:
Is it necessary to scale the std by `numSamples` for `labelStd`, `featureStd` and `covariance`? Mathematically it's the same without `numSamples` for these three variable. However, it might affect the choice of `MathConst.EPSILON` for stably computing the score. Say `numSamples` is 100M, then the product of `labelStd * featureStd` will be very small. I meant it's not in the same magnitude of the `denominator` in the current code.

Reply:
I'm not sure what you suggest, but here what I think for very large `numSamples` (let's say 100M as you suggested).

Before normalization, the three "unscaled" variables are sums of 100M terms with mean 0. The rescaling brings these very large sum back to the scale of the data's variance / std.

Even if the rescaling can be skipped, it is necessary to rescale the epsilon at two different place, so it doesn't really make a difference in term of code.

I wouldn't mind implementing what you suggest, but I'm not sure which benefits there would be. I might simply be missing something though :)

Reply:
I might misunderstand something from your comment, but assuming we have a dataset of 100M observations:

1. Removing the  `numSamples` normalization from the var and std will require - as you mentioned - to rescale `math.epsilon`. This means we wouldn't have that much of a cosmetic advantage of doing that.

2. The variables you mention are sums of 100M positive numbers centered around 0, they are potentially very big - same order of magnitude as `numSamples`. Normalizing them brings them back to the scale of the var and std, so i'm not sure why the product you mentioned should be that small.

I might be missing something obvious, otherwise I'm not sure what could motivate this change.

Reply:
Don't worry this is a minor issue which won't affect the result too much. My concern is that since the old code did not normalize the numerator and denominator, we might need to be careful about choosing the EPSILON.

Reply:
> @matthieubulte @cmjiang
> My preference would be similar to option 2, and exactly what Chengming suggested: rename this function to `stableComputePearsonCorrelationScore` and add another unit test with a `SparseVector`. Let's keep both functions around for now, as it's unclear what the future of feature filtering using Pearson correlation score will be.
> 
> Thank you very much for the contribution @matthieubulte

I will take care of this as well.